### PR TITLE
add explanations and examples to the docstring of violinplot

### LINKED ISSUE #12178: Violin plot inconsistency between list and numpy
### Bug report

When feeding the same data to violin plot in list or in numpy array, the result is not the same. 

**Code for reproduction**

```python
import matplotlib.pyplot as plt
import numpy as np

data = [ [1, 2, 3] , [4, 5, 6] ]
data_np = np.array(data)

f, (ax1, ax2) = plt.subplots(2)

ax1.violinplot(data)
ax1.set_title('list')

ax2.set_title('numpy array')
ax2.violinplot(data_np) #Same data but in numpy

plt.show()

```

**Actual outcome**

problem

**Expected outcome**
Both should be the same isn't it ?

Actually the [doc](https://matplotlib.org/api/_as_gen/matplotlib.pyplot.violinplot.html#matplotlib.pyplot.violinplot) specify - in terms I didn't understood at first - that it will
"Make a violin plot for each column of dataset or each vector in sequence dataset. " 

It is more clear when looking at the [hist doc](https://matplotlib.org/api/_as_gen/matplotlib.pyplot.hist.html), where one can read at the end

"Note that the ndarray form is transposed relative to the list form."

The difference between the two outcomes, I think, is the way of thinking at data : vectors in matrix or lists in list (which correspond to columns in array of row in array). 
I discussed with some collegues who are working with Matlab, and they do think the vector as base unit of the 2D matrix. For me the base unit is a row of a 2D array.  

In the function plot when plotting [ [1,2,3], [4,5,6]  ], we have 3 curves, so I thing it is the matrix/vector way of thinking that is predominant in matplotlib. 
So should we change the violinplot and histogram behavior for list to work the same as array ?

After writing this I do belive that the best solution is to just mention that specificity more clearly in the doc of violinplot, not changing the code. I can do this, but I would like some feedback from differents points of view before.

Best,
RP

### PR DESCRIPTION
**Pull request by andrea-w-wang**
Add explanations and examples to the docstring of violinplot to explain how violinplot handles differently w.r.t. sequence of vectors and 2D array. 
Closes #12178

## PR Checklist

- [V] Has Pytest style unit tests
- [V] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [V] New features are documented, with examples if plot related
- [V] Documentation is sphinx and numpydoc compliant
- [V] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [V] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -7811,14 +7811,19 @@ def violinplot(self, dataset, positions=None, vert=True, widths=0.5,
...
+        Each filled area extends to represent the entire data range, with optional lines at the mean, the median,
         the minimum, and the maximum.
 
         Parameters
         ----------
-        dataset : Array or a sequence of vectors.
-          The input data.
+        dataset : 2D Array/Matrix or a sequence of vectors.
+          The input data. 
+          If passing a 2D Array/Matrix, such as *np.array([[1,2,3], [4,5,6]])*, it will plot (1,4), (2,5), and (3,6)
```
Comment (story645) [inline lib/matplotlib/axes/_axes.py]: Thinking this should either be in a  note or directly in the example.

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -7811,14 +7811,19 @@ def violinplot(self, dataset, positions=None, vert=True, widths=0.5,
         Make a violin plot.
 
         Make a violin plot for each column of *dataset* or each vector in
-        sequence *dataset*.  Each filled area extends to represent the
-        entire data range, with optional lines at the mean, the median,
+        sequence *dataset*.  To illustrate, if passing an 2D array/matrix,
```
Comment (story645) [inline lib/matplotlib/axes/_axes.py]: I don't think it needs the to illustrate if you put this same information in the note.

**Comment by story645** [comment]
Hi @wywang-nyu, this PR is very welcome and thank you for making it. Do you plan on making the requested changes?

**Comment by wywang-nyu** [comment]
> Hi @wywang-nyu, this PR is very welcome and thank you for making it. Do you plan on making the requested changes?

Hi, sorry for the delay! I totally forgot about this. I just made the request changes, please review! Thank you.

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -7811,14 +7811,15 @@ def violinplot(self, dataset, positions=None, vert=True, widths=0.5,
...
-        entire data range, with optional lines at the mean, the median,
+        sequence *dataset*. Each filled area extends to represent the entire data range, with optional lines at the mean, the median,
         the minimum, and the maximum.
 
         Parameters
         ----------
-        dataset : Array or a sequence of vectors.
-          The input data.
+        dataset : 2D Array/Matrix or a sequence of vectors.
+          The input data.
```
Comment (story645) [inline lib/matplotlib/axes/_axes.py]: don't need "the input data".  Notes about data haven't been moved to notes section"

**Comment by story645** [comment]
closing because this got kind of addressed in #14107 but will totally reopen if @andrea-w-wang has the bandwidth to readdress this.

**Pull request closed without merging**